gluten-core/src/main/scala/io/glutenproject/extension/columnar/ColumnarPullOutPreProject.scala

```diff
@@ -0,0 +1,117 @@
+ * attributes. To ensure that the downstream post-project can obtain the correct outputOrdering, it
+ * is necessary to restore the original sortOrder with this tag hint Otherwise, when verifying
+ * whether the output attributes of the downstream are a subset of the references of
+ * sortOrder.child, it will not pass. Please check AliasAwareQueryOutputOrdering.outputOrdering in
+ * Spark-3.4.
+ */
+object SortOrderHint {
+  val TAG: TreeNodeTag[Seq[SortOrder]] =
+    TreeNodeTag[Seq[SortOrder]]("io.glutenproject.sortorderhint")
+
+  def getSortOrder(plan: SparkPlan): Option[Seq[SortOrder]] = {
+    plan.getTagValue(TAG)
+  }
+
+  def tag(plan: SparkPlan, sortOrder: Seq[SortOrder]): Unit = {
+    plan.setTagValue(TAG, sortOrder)
+  }
+}
+
+/**
+ * The native engine only supports executing Expressions within the project operator. When there are
+ * Expressions within physical operators such as SortExec and HashAggregateExec in SparkPlan, we
+ * need to pull out the Expressions to be executed earlier in the pre-project operator. This rule is
+ * to transform the SparkPlan at the physical plan level, constructing a SparkPlan that supports
+ * execution by the native engine.
+ */
+case class ColumnarPullOutPreProject(session: SparkSession)
+  extends Rule[SparkPlan]
+  with PullOutProjectHelper {
+
+  private def needsPreProject(plan: SparkPlan): Boolean = {
+    if (notSupportTransform(plan)) {
+      // TagBeforeTransformHits may tag the plan as not transformable for some reason,
+      // for example, when ansi mode is enabled. In such cases, there is no need to
+      // pull out the project.
+      return false
+    }
+    plan match {
+      case sort: SortExec =>
+        sort.sortOrder.exists(o => isNotAttribute(o.child))
+      case take: TakeOrderedAndProjectExec =>
+        take.sortOrder.exists(o => isNotAttribute(o.child))
+      case _ => false
+    }
+  }
+
+  override def apply(plan: SparkPlan): SparkPlan = PhysicalPlanSelector.maybe(session, plan) {
+    plan.transform {
+      case sort: SortExec if needsPreProject(sort) =>
+        val expressionMap = new mutable.HashMap[Expression, NamedExpression]()
+        // The output of the sort operator is the same as the output of the child, therefore it
+        // is necessary to retain the output columns of the child in the pre-projection, and
+        // then add the expressions that need to be evaluated in the sortOrder. Finally, in the
+        // post-projection, the additional columns need to be removed, leaving only the original
+        // output of the child.
+        val newSortOrder =
+          sort.sortOrder.map(_.mapChildren(replaceExpressionWithAttribute(_, expressionMap)))
+
+        val preProject = ProjectExec(
+          eliminateProjectList(sort.child.output, expressionMap.values.toSeq),
+          sort.child)
+
+        val newSort =
+          sort.copy(sortOrder = newSortOrder.asInstanceOf[Seq[SortOrder]], child = preProject)
```

Comment:
Can we add the original sort order into the sameOrderExpressions ? e.g., the original orders are `Seq(expr(c1), c2)`, and newOrders are `Seq(_pre_1, c2)`. We can zip these two orders and inject new into original sameOrderExpressions if the order is different. So we do not need tag any more.

Reply:
Good idea. I will test later.